### PR TITLE
CHE-5684: implement URLRewriter for MacOS

### LINKED ISSUE #5684: Implement URLRewriter for Docker infrastructure
Related to SPI branch only.
We replaced the concept of ServerEvaluationStrategy with URLRewriter in spi branch. Now URLRewriter impl doesn't change URLs at all. We need to create impl that can cover Docker infra needs, such as local Linux, remote Linux,  Windows, Mac environments.

### PR DESCRIPTION
**Pull request by garagatyi**
### What does this PR do?
Implement and bind as default impl URLRewriter which makes Docker
infra work on MacOS or Windows. Property 'che.docker.ip.external'
should be used to set the correct external hostname for containers.
Refactored URLRewriter interface to use strings as URLs instead of
URL class since it can't handle non-basic protocols, e.g. ws.

### What issues does this PR fix or reference?
Fixes #5684

### CODE REVIEW

**Review by riuvshin — APPROVED** [review]

File: infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/URLRewriterImpl.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import static java.lang.String.format;
+
+/**
+ * {@link URLRewriter} needed for {@link DockerRuntimeInfrastructure} to handle specific cases of
+ * running containers.
+ * </br> For example, containers in MacOS, Windows, etc.
+ *
+ * @author Alexander Garagatyi
+ */
+public class URLRewriterImpl implements URLRewriter {
```
Comment (voievodin) [inline infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/URLRewriterImpl.java]: Maybe change the name to something more certain like `ExternalIpURLRewriter` ?

Comment (garagatyi) [reply infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/URLRewriterImpl.java]: ok

**Review by skabashnyuk — APPROVED** [review]

**Comment by benoitf** [comment]
I don't see how we can plug additional url rewriters like when we're using a reverse proxy with that new way. 

Also this URL rewriter is only handling che.docker.ip.external while we had possibility to select by property the custom strategy pattern in Che 5.

How it is handle in this branch ? as I mainly see it as a regression.

**Comment by garagatyi** [comment]
To bind another Url Rewriter you need to replace url rewriter class in constructor with interface of url rewriter and bind it. 

This PR doesn't handle strategies implementation. If it is needed it should be implemented separately.

**Comment by benoitf** [comment]
@garagatyi yes but this is still not configurable as it's hardcoded in the class.
What I see is that each time a new design is bring, each time you drop all features that I added before. This is why  I want to warn it into this PR as it's again the case.
AFAIK SPI should bring new design but keep existing features.

**Review by gazarenkov — COMMENTED** [review]
@garagatyi can we use uri intead of url (you replaced with string) ?

@benoitf che6 is not backward compatible

**Comment by benoitf** [comment]
@gazarenkov I know it's not backward compatible but architecture design should still take into account previous features which are useful for end users.

**Comment by garagatyi** [comment]
> yes but this is still not configurable as it's hardcoded in the class

Since there are no other strategies it doesn't make sense to implement strategy selection. 

> can we use uri intead of url (you replaced with string) ?

@gazarenkov  we can, but here is the [only place where it is used](https://github.com/eclipse/che/blob/spi/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/InternalRuntime.java#L201) and it is converted into String right there. So even if we change it to URI this fact won't change anything at all.

**Review by sleshchenko — APPROVED** [review]

**Review by akorneta — APPROVED** [review]

**Review by voievodin — APPROVED** [review]

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/InternalRuntime.java [hunk 2]
```diff
@@ -197,22 +197,11 @@ public final T getContext() {
...
-                        return 101;
-                    }
-
-                    @Override
-                    public String getMessage() {
-                        return "Malformed URL for " + name + " : " + e.getMessage();
-                    }
-                });
+            } catch (InfrastructureException e) {
+                warnings.add(new WarningImpl(101, "Malformed URL for " + name + " : " + e.getMessage()));
```
Comment (voievodin) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/InternalRuntime.java:204]: What is _101_?
Have you considered implementing warning for this certain case e.g.
```java
class MalformedURLWarning implements Warning { 
  static int CODE = 101;
  //...
  int getCode() { return CODE; }
  //...
}
```

Comment (garagatyi) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/spi/InternalRuntime.java:204]: it is initial value  - I didn't change it. I suppose we need to think about spec on it and then do something like class you have suggested

**Review by gazarenkov — APPROVED** [review]
OK, 
but let's consider (discuss) replace String with URI in all related classes

**Pull request merged**